### PR TITLE
Descriptor configs

### PR DESCRIPTION
**Pull request by jmosbacher**
This started as an alternative solution to XENONnT/straxen#684, but has expanded in scope to facilitate a CMT upgrade. the problem is that CMT logic was creeping into the plugin code written by analysts, often being duplicated in multiple plugins.

This PR implements the descriptor protocol for config options so they can run arbitrary code when accessed. Config values can then be accessed like a regular attribute from inside the plugin logic. 

A few reference implementations of configs types are added to the documentation:

- CallableConfig : calls a user given function with the dependencies as arguments and returns the value
- LookupConfig: used a user supplied mapping to lookup the value based on the dependency values.
- RemoteConfig: goes through a list of user supplied storages and calls .get_value(name, **kwargs) on each until a value is returned which is not None.
-  URLConfig: if config value is a string, it parses as a URL and dispatches a registered function based on the protocol/scheme part of the URL. Passes the path part of the URL as argument and the url args as kwargs to the dispatched function, where the plugin attributes can be accessed at runtime and passed as kwargs. Protocols can be chained to achieve complex lookups set by the URL. 

The URLConfig is the most flexible and advanced example and can essentially replace all the others but in most projects such flexibility is not needed. functions can be registered to it by using the URLConfig.register(`NAME`) decorator. these functions are called when the config is set to a url defining the protocols to dispatch e.g. 'cmt://corection_name?version=1&run_id=plugin.run_id'. its also possible to chain protocols e.g. 'unzip://cmt://corection_name?version=1&run_id=plugin.run_id' in which case the outer protocol will be called with the result of the inner protocol chain.

Example:
```python
import strax


# Register some protocols
@strax.URLConfig.register('cmt')
def get_correction(name, **kwargs):
    run_id = kwargs.get('run_id', None)
    if run_id is None:
        raise
    # value = ...  fetch the config value from CMT using the name and kwargs
    return  f'{name} was fetched from cmt with run_id: {run_id}'

@strax.URLConfig.register('straxen')
def get_resource(name, **kwargs):
    # value = ...  fetch the config value e.g. from the straxen file downloader
    return f'{name} was fetched from straxen '

@strax.URLConfig.register('fsspec')
def read_file(path, **kwargs):
    with fsspec.open(path, **kwargs) as f:
        content = f.read()
    return content

@strax.URLConfig.register('json')
def read_json(content, **kwargs):
    return json.loads(content)


class DummyPlugin(strax.Plugin):
    depends_on = ()
    first_config = strax.URLConfig()
    second_config = strax.URLConfig()
   
# set config values to URLs using the registered protocols to define access
p = DummyPlugin()
p.config = {
             'first_config': 123,
             'second_config': 'cmt://correction_name?version=v1&run_id=plugin.run_id',
}
p.run_id = 321

# accessing first_config will return 123
print(p.first_config)
>>> 123

# accessing second_config will call `get_correction` function
print(p.second_config) 
>>> ...

p.config = {
          'first_config': 'json://fsspec://file://test.json',
          'second_config': 'straxen://some_file?version=v1',
}


print(p.first_config) # will first call `read_file` then pass the result to `read_json`

print(p.second_config) # will call `get_resource` function


```

Some more examples:

```python


def modify_the_value(value, run_id, simple_config):
    #do some magic here
    return value

value_by_run_id = {'run_0': 0, 'run_1':1}

class DummyPlugin(strax.Plugin):
    '''define some dependencies if you want to run this'''
    function_config = strax.CallableConfig(func=modify_the_values, args=('run_id', 'simple_config'))
    lookup_config = strax.LookupConfig(value_by_run_id, keys=['run_id'])
    
    def compute(self, **kwargs):
         print(f"the value of function_config is {self.function_config}")

```

when plugin.function_config is accessed, the value is retrieved from the plugin config and then the modifier is called with the value in plugin.config(the one being hashed), run_id and the value of plugin.simple_config, the result is then returned.

### CODE REVIEW

**Comment by jmosbacher** [comment]
A straxen example using CMT:

```python
@export
@strax.takes_config(
    strax.Option('min_reconstruction_area',
                 help='Skip reconstruction if area (PE) is less than this',
                 default=10),
    strax.Option('n_top_pmts', default=straxen.n_top_pmts,
                 help="Number of top PMTs")
)
class PeakPositionsBaseNT(strax.Plugin):

    def setup(self):
        self.model_file = self._get_model_file_name()
        if self.model_file is None:
            warn(f'No file provided for {self.algorithm}. Setting all values '
                 f'for {self.provides} to None.')
            # No further setup required
            return

        # Load the tensorflow model
        import tensorflow as tf
        if os.path.exists(self.model_file):
            print(f"Path is local. Loading {self.algorithm} TF model locally "
                  f"from disk.")
        else:
            downloader = straxen.MongoDownloader()
            try:
                self.model_file = downloader.download_single(self.model_file)
            except straxen.mongo_storage.CouldNotLoadError as e:
                raise RuntimeError(f'Model files {self.model_file} is not found') from e
        with tempfile.TemporaryDirectory() as tmpdirname:
            tar = tarfile.open(self.model_file, mode="r:gz")
            tar.extractall(path=tmpdirname)
            self.model = tf.keras.models.load_model(tmpdirname)
            
    def _get_model_file_name(self):
        config_file = f'{self.algorithm}_model'
        model_from_config = self.config.get(config_file, 'No file')
        if model_from_config == 'No file':
            raise ValueError(f'{__class__.__name__} should have {config_file} '
                             f'provided as an option.')
        if isinstance(model_from_config, str) and os.path.exists(model_from_config):
            # Allow direct path specification
            return model_from_config
        if model_from_config is None:
            # Allow None to be specified (disables processing for given posrec)
            return model_from_config

        # Use CMT
        model_file = straxen.get_correction_from_cmt(self.run_id, model_from_config)
        return model_file

```
Notice how all the details on how to fetch the model file and convert it to a python object to be using is hardcoded into the plugin. This is not desirable, the plugin should contain processing logic only.

How this would be refactored using `strax.URLConfig`:

```python
# ------------------ These protocols would be defined in straxen for use with all plugins/ at the beginning of the script/notebook when doing custom stuff.
@strax.URLConfig.register('cmt')
def get_correction(name, run_id=0, version='ONLINE', detector='nt', **kwargs):
    return straxen.get_correction_from_cmt(run_id, ( name, version, detector=='nt'))

@strax.URLConfig.register('resource')
def get_resource(name, fmt='text', **kwargs):
    return straxen.get_resource(name, fmt=fmt)

@strax.URLConfig.register('format')
def format_arg(arg, **kwargs):
    return arg.format(**kwargs)
    
@strax.URLConfig.register('keras')
def load_keras_model(content, **kwargs):
    fd = io.BytesIO(content)
    with tempfile.TemporaryDirectory() as tmpdirname:
        tar = tarfile.open(fileobj=fd, mode="r:gz")
        tar.extractall(path=tmpdirname)
        model = tf.keras.models.load_model(tmpdirname)
    fd.close()
    return model
# ------------------------------------------------------------------------

"""
Then the plugin would just declare that it needs a keras model to function, with no details on where the model comes from hardcoded in the plugin itself.
"""
class PeakPositionsBaseNT(strax.Plugin):
    model = strax.URLConfig(type=tf.keras.Model, cache=True)

# in the context:
config = {
    'model': 'keras://resource://cmt://format://{algorithm}_model?version=v1&run_id=plugin.run_id&fmt=binary&algorithm=plugin.algorithm',
}


# When plugin.model is accessed by the compute logic, the algorithm attribute of the plugin will be passed to the `format_arg` function, the resulting correction name will be passed to the `get_correction` function with the run_id and version as keyword arguments, which will return a filename that will be passed to the `get_resource` function along with the fmt=binary keyword argument, the resulting bytestring will be passes to the `load_keras_model` function that will return the model for use. since loading a model is usually an expensive operation we enable caching.
```

File: strax/config.py [hunk 1]
```diff
@@ -203,6 +218,224 @@ def validate(self, config,
...
+            else:
+                v = k
+            key.append(v)
+        if len(key)==1:
+            key = key[0]
+        else:
+            key = tuple(key)
+        return self.mapping[key]
+@export
+class RemoteConfig(Config):
```
Comment (JoranAngevaare) [inline strax/config.py:289]: I can imagine some slight issues concerning reproducibility with this class, if a value can come from any storage, and it's up to the user to make sure that either the storage's all provide unique names or that they always return the same. I know that to some extend this can also easily be argued for the other Options, but at least there you know more or less the hopefully unique path to the requested value. We don't track which storage is providing the requested information, this might both be a feature or a bug, depending fully on the setup one chooses for 😉.

File: strax/config.py [hunk 2]
```diff
@@ -203,6 +218,224 @@ def validate(self, config,
...
+        key = []
+        for k in self.keys:
+            if k=='name':
+                v = self.name
+            elif k=='value':
+                v = plugin.config[self.name]
+            elif isinstance(k, str) and hasattr(plugin, k):
+                v = getattr(plugin, k)
+            else:
+                v = k
```
Comment (JoranAngevaare) [inline strax/config.py:281]: at this point you lost me, maybe reviewing at this time also isn't the greatest idea but this is supposed to just return the key itself as the value?

**Review by JoranAngevaare — COMMENTED** [review]
Thanks a lot Yossi, I think it's an awesome piece of code you wrote. I especially like the `URLConfig`. Especially the chaining of protocols opens the path to unlimited combinations and generalizations while still providing a more or less unique roadmap to the value.

I do have some (sometimes pedantic) comments (as below). 

You added quite a lot of options but that also makes it sometimes a bit hard to see their disguising use-cases, especially given the already very  general `URLConfig`. If we add all these different classes, I would argue for documenting their use cases well and making their differences clear, or we could choose the "best" ones and focus on those. Just to prevent adding code that no-one uses or isn't quite sure of what it's best used for. What do you think is the best approach?

Here and there, the code is a bit tars (short notation) and/or can do with a few comments more. As Chris likes to say, code is read more than it's written 😉.

File: tests/test_config.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+def test_lookup_config():
+    p = DummyPlugin()
+    p.run_id = 321
+
+    assert p.yay == 'yay!'
+
+    p.run_id = 123
+
+    assert p.yay == 'nay!'
+
```
[the remark this answers is not in the document]
Comment (JoranAngevaare) [reply tests/test_config.py:36]: like badly formatted urls, what do we want to user to see and do we make sure the right (helpfull) errors are raised

**Comment by jmosbacher** [comment]
@JoranAngevaare thank you for the great comments, its getting late so will fix them all tomorrow but just wanted to propose for the multiple "reference implementations" i made here to demonstrate how to subclass the Config class, how about i move them out of the code and into the docs? so ill put them in the docs as examples of things you can do by subclassing Config but not have them actually in the code. What do you think? I guess i can just leave the URLConfig or maybe just move it to straxen and only leave the Config class in strax?

File: strax/config.py [hunk 4]
```diff
@@ -203,6 +218,224 @@ def validate(self, config,
...
+    def register(cls, protocol, func=None):
+        """Register dispatch of `func` on urls
+         starting with protocol name `protocol` """
+
+        def wrapper(func):
+            if isinstance(protocol, tuple):
+                for t in protocol:
+                    cls.register(t, func)
+            else:
+                cls._lookup[protocol] = func
```
[the remark this answers is not in the document]
Comment (jmosbacher) [reply strax/config.py:381]: Yeah i was very much on the fence with that. Im really not sure what the best behavior would be. Maybe just a warning that its being overridden would be enough?

File: strax/config.py [hunk 5]
```diff
@@ -203,6 +218,224 @@ def validate(self, config,
...
+@export
+class Config(Option):
+    def __init__(self, **kwargs):
+        if 'name' not in kwargs:
+            kwargs['name'] = ''
+        super().__init__(**kwargs)
+
+    def __set_name__(self, owner, name):
+        self.name = name
+        takes_config = {name: self}
```
[the remark this answers is not in the document]
Comment (jmosbacher) [reply strax/config.py:231]: Not sure i understand what you mean here, is the issue the readability or the functionality. this code performs the tasks that the takes_config decorator would perform on class instantiation. Since I dont use the decorator I have no need for it in this namespace and it seamed like the most descriptive name for what it holds, is it more confusing than descriptive?

File: strax/config.py (hunk 2 above)
Comment (jmosbacher) [reply strax/config.py:280]: haha yeah i probably also wrote that sometime during the night... it allows for "constant" keys. So if I have a dict-like object where one of the keys needs to just always be some constant then I can pass that constant in the keys list and it will always  be included in the lookup key...

**Comment by WenzDaniel** [comment]
Maybe it is still to early, but I must admit that for me it is very hard to follow. Although I really like the flexibility you get with descriptors and url styled configs I am a little bit afraid now that things are now more complicated than with the stuff we started with. I think you got a bit carried away. While in principle flexibility is good, I think in this particular case too much flexibility can do more harm than good as things are getting more complicated and messier. 

I am especially afraid of

```
    simple_config = strax.Config(default=1, type=int)
    function_config = strax.CallableConfig(func=modify_the_values, args=('run_id', 'simple_config'))
    lookup_config = strax.LookupConfig(value_by_run_id, keys=['run_id'])
    remote_config = strax.RemoteConfig(storages)
    cmt_config = strax.URLConfig()
```

Does this mean that every different config type got now there own unique way to specify options? Do I have to remember now as developer n different methods for specifying an option?

**Comment by JoranAngevaare** [comment]
> 
> 
> @JoranAngevaare thank you for the great comments, its getting late so will fix them all tomorrow but just wanted to propose for the multiple "reference implementations" i made here to demonstrate how to subclass the Config class, how about i move them out of the code and into the docs? so ill put them in the docs as examples of things you can do by subclassing Config but not have them actually in the code. What do you think? I guess i can just leave the URLConfig or maybe just move it to straxen and only leave the Config class in strax?

@jmosbacher That sounds like a good compromise to put them in the developer docs as far as I am concerned, thanks!

File: strax/config.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (JoranAngevaare) [reply strax/config.py:231]: For me it was just readability

File: strax/config.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (JoranAngevaare) [reply strax/config.py:380]: I would argue that an ValueError would not be misplaced. Or make an option to let it overwrite (e.g. during development where you don't want to manually de register maybe).

A warning would also work, I think this is one of the cases where a slow transition will learn us what is practically best and where there are many ways leading to Rome.

At least silently overwriting (at this stage) seems to me a place where things can get messy. If they happen in  import straxen`, I don't know what was registered last

**Comment by jmosbacher** [comment]
@WenzDaniel @JoranAngevaare thanks for the very important feedback. I think you both hit on a very important point that I overlooked. The reference implementations of the Config subclasses give the impression that they should be used together in the same project. This of course is probably a very bad idea for most projects, especially in projects like straxen which involve many collaborators that are not the plugin developers who would need to try to figure out what syntax to use for each config option. I have moved all reference implementations to the plugin_dev section of the documentation and will open a PR to make the URLConfig implementation a part of straxen. I think this would be the cleanest approach, I hope you agree.

File: docs/source/advanced/plugin_dev.rst [hunk 6]
```diff
@@ -43,6 +43,315 @@ You can specify defaults in several ways:
...
+                if isinstance(v, str) and hasattr(plugin, v):
+                    kwargs[k] = getattr(plugin, v)
+                else:
+                    kwargs[k] = v
+            
+            value = super().fetch(plugin)
+            value = self.func(value, *args, **kwargs)
+            return value
+
+URL style configuration
```
Comment (JoranAngevaare) [inline docs/source/advanced/plugin_dev.rst:202]: maybe add a link to straxen?

**Review by JoranAngevaare — APPROVED** [review]
Thanks a lot Yossi!

**Review by WenzDaniel — APPROVED** [review]
Looks fine to me. Thanks a lot for this great work. It is still a bit abstract for me but I am already looking forward to play with it.

**Comment by JoranAngevaare** [comment]
Thanks Yossi!

**Review by JoranAngevaare — APPROVED** [review]
